system/lib/wasmfs/emscripten.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2023 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+//
+// This file contains implementations of emscripten APIs that are compatible
+// with WasmFS. These replace APIs in src/library*js, and basically do things
+// in a simpler manner for the situation where the FS is in wasm and not JS
+// (in particular, these implemenations avoid calling from JS to wasm, and
+// dependency issues that arise from that).
+//
+
+#include <emscripten.h>
+#include <stdio.h>
+
+#include <string>
+
+#include "file.h"
+#include "file_table.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+// Given an fd, returns the string path that the fd refers to.
+// TODO: full error handling
+// TODO: maybe make this a public API, as it is useful for debugging
+std::string getPath(int fd) {
```

Comment:
Why not use this to implement the existing `FS.getStream()`?

Reply:
I'd really rather us implement as little as possible in these semi-internal FS APIs. By "semi-internal" I mean that `FS.open` for example is an impl of libc `open`, while `FS.getStream()` is just an internal API that the JS FS had. Some of those APIs, like `getStream` and `analyzePath` return objects with lots of interesting internal properties.

We can and should "polyfill" the "public" APIs like `open`, I think, since users might be using them. But for internal usage of the filesystem I think the natural thing is to move code from JS to wasm. Eventually, we'll get rid of the JS FS and then it will all be in one place (the proper one). That's my long-term goal here. Does that make sense?